Tasks/CargoAuthenticateV0/cargoauthenticatemain.ts

```diff
@@ -0,0 +1,72 @@
+import * as path from 'path';
+import * as tl from 'azure-pipelines-task-lib/task';
+import * as toml from 'toml';
+import * as fs from 'fs';
+import * as constants from './constants';
+import * as util from 'azure-pipelines-tasks-packaging-common/util';
+import * as pkgLocationUtils from 'azure-pipelines-tasks-packaging-common/locationUtilities';
+import * as url from 'url';
+
+async function main(): Promise<void> {
+    tl.setResourcePath(path.join(__dirname, 'task.json'));
+
+    try {
+        let configtoml = tl.getInput(constants.CargoAuthenticateTaskInput.ConfigFile);
+        if (!tl.exist(configtoml)) {
+            throw new Error(tl.loc('ConfigTomlDoesNotExist', configtoml));
+        }
+        else {
+            console.log(tl.loc("AuthenticatingThisConfigToml", configtoml));
+        }
+        
+        // These two formats will work
+        // [registries]
+        // zhentan-test = { index = "sparse+https://pkgs.dev.azure.com/codesharing-SU0/zhentan-test/_packaging/zhentan-test/Cargo/index/" }
+        // [registries.zhentan-test1]
+        // index = "sparse+https://pkgs.dev.azure.com/codesharing-SU0/zhentan-test/_packaging/zhentan-test/Cargo/index/"
+        let configtomlFile = fs.readFileSync(configtoml, 'utf8');
+        var result = toml.parse(configtomlFile);
+        if (!result.registries)
+        {
+            throw new Error(tl.loc('ConfigTomlMissingRegistry'));
+        }
+
+        let packagingLocation: pkgLocationUtils.PackagingLocation;
+        try {
+            // TODO: Change Npm to Cargo once new packaging common is deployed
+            packagingLocation = await pkgLocationUtils.getPackagingUris(pkgLocationUtils.ProtocolType.Npm);
+        } catch (error) {
+            tl.debug('Unable to get packaging URIs');
+            util.logError(error);
+            throw error;
+        }
+
+        const collectionHosts = packagingLocation.PackagingUris.map((pkgUrl: string) => {
+            const parsedUrl = url.parse(pkgUrl);
+            if (parsedUrl && parsedUrl.host) {
+                return parsedUrl.host.toLowerCase();
+            }
+            return undefined;
+        });
+
+        const localAccesstoken = `Bearer ${tl.getVariable('System.AccessToken')}`;
+        for (let registry of Object.keys(result.registries)) {
+            const registryUrl = url.parse(result.registries[registry].index);
+            if(registryUrl && registryUrl.host && collectionHosts.indexOf(registryUrl.host.toLowerCase()) >= 0) {
```

Comment:
Does this work with both old and new style urls (`dev.azure.com/<collection>` and `<collection>.visualstudio.com`) ?

Reply:
Yeah, collectionHosts will contain all URL that we support. We do the same validation for npm since we allow user to pass in registry URL in .npmrc 

Example pipeline debug log, we just compare the host we have against the host parsed from the URL provided by the user 
![image](https://user-images.githubusercontent.com/32987404/205147757-35767db6-aab7-4e98-a1b5-47f8ce1a6d16.png)